ofrak_core/ofrak/gui/script_builder.py

```diff
@@ -0,0 +1,297 @@
+    """
+
+    def __init__(self):
+        self.root_cache: Dict[bytes, Resource] = {}
+        self.script_sessions: Dict[bytes, ScriptSession] = {}
+        self.selectable_indexes: List[ResourceIndexedAttribute] = [
+            FilesystemEntry.Name,
+            Data.Offset,
+        ]
+
+    async def add_action(
+        self,
+        resource: Resource,
+        action: str,
+        action_type: ActionType,
+    ) -> None:
+        """
+        :param action:
+        :param action_type:
+        """
+        var_name = await self.add_variable(resource)
+        qualified_action = action.format(resource=var_name)
+        await self._add_action_to_session(resource, qualified_action, action_type)
+
+    async def add_variable(self, resource: Resource) -> bytes:
+        if await self._var_exists(resource):
+            return await self._get_variable_from_session(resource)
+
+        root_resource = await self._get_root_resource(resource)
+        if resource.get_id() == root_resource.get_id():
+            await self._add_variable_to_session(resource, "root_resource")
+            return "root_resource"
+
+        parent = await resource.get_parent()
+        if not await self._var_exists(parent):
+            await self.add_variable(parent)
+
+        selector = await self._get_selector(resource)
+        name = await self._generate_name(resource)
+        await self._add_action_to_session(
+            resource,
+            rf"""
+        {name} = {selector}""",
+            ActionType.UNDEF,
+        )
+        await self._add_variable_to_session(resource, name)
+        return name
+
+    async def delete_action(self, resource: Resource, action: str) -> None:
+        """
+        :param action:
+        """
+        root_resource = await self._get_root_resource(resource)
+        session = self._get_session(root_resource.get_id())
+        # TODO: do we really need to delete an action from the script?
+        for idx, script_action in enumerate(session.actions):
+            if script_action.action == action:
+                session.actions.pop(idx)
+
+    async def get_script(self, resource: Resource) -> str:
+        """
+        :return script:
+        """
+        root_resource = await self._get_root_resource(resource)
+        return self._get_script(root_resource.get_id())
+
+    async def get_all_of_type(self, resource: Resource, target_type: ActionType) -> str:
+        """
+        :param target_type:
+        :return script:
+        """
+        root_resource = await self._get_root_resource(resource)
+        return self._get_script(root_resource.get_id(), target_type)
+
+    async def _get_root_resource(self, resource: Resource) -> Resource:
+        if resource.get_id() in self.root_cache:
+            return self.root_cache[resource.get_id()]
+        while len(list(await resource.get_ancestors())) != 0:
+            resource = await resource.get_parent()
+        self.root_cache[resource.get_id()] = resource
```

Comment:
This does not properly update the cache. It uses the ID of the root and sets its value to the root. The ID should be stored in a variable before the destructive loop above.

At the moment, the cache is effectively unused.